File: src/tools/validate-citation.ts
/**
 * validate_citation — Validate an Hungarian legal citation against the database.
 */

import type Database from '@ansvar/mcp-sqlite';
import { resolveDocumentId } from '../utils/statute-id.js';
import { generateResponseMetadata, type ToolResponse } from '../utils/metadata.js';

export interface ValidateCitationInput {
  citation: string;
}

interface ValidateCitationResult {
  valid: boolean;
  citation: string;
  normalized?: string;
  document_id?: string;
  document_title?: string;
  provision_ref?: string;
  status?: string;
  warnings: string[];
}

/**
 * Parse an Hungarian legal citation.
 * Supports:
 * - Hungarian formal: "2012. évi I. törvény 116. §"
 * - Database ID: "hu-law-2012-1-00-00 s116"
 * - "Section 3 Infotörvény" / "Section 3, Infotörvény"
 * - "Infotörvény s 3" / "Infotörvény, s 3"
 * - "[Act Title Year] s N"
 * - "s 13" (section only, no document)
 * - Plain document reference (e.g., "Infotörvény")
 */
const SECTION_FIRST_RE = /^Section\s+(\d+[A-Za-z]*(?:\(\d+\))?)\s*[,;]?\s+(.+)$/i;
const SECTION_LAST_RE = /^(.+?)\s*[,;]?\s+(?:s\.?\s+|Section\s+)(\d+[A-Za-z]*(?:\(\d+\))?)$/i;

export interface ParsedCitation {
  documentRef: string;
  sectionRef?: string;
  /** True when documentRef is a formal Hungarian reference or a database ID. */
  structured: boolean;
}

export function parseCitation(citation: string): ParsedCitation | null {
  const trimmed = citation.trim();
  if (!trimmed) return null;

  // Hungarian formal: "2012. évi I. törvény 116. §" or "2013. évi V. törvény 6:272. §" or "116/A. §"
  const hungarianFull = trimmed.match(
    /^(\d{4}\.\s*évi\s+[IVXLCDM]+\.\s*törvény)\s+(\d+(?::\d+)?(?:\/[A-Za-z])?)\.\s*§/i
  );
  if (hungarianFull) {
    return { documentRef: hungarianFull[1].trim(), sectionRef: hungarianFull[2], structured: true };
  }

  // Hungarian document only: "2012. évi I. törvény" (no section)
  const hungarianDoc = trimmed.match(
    /^(\d{4}\.\s*évi\s+[IVXLCDM]+\.\s*törvény)$/i
  );
  if (hungarianDoc) {
    return { documentRef: hungarianDoc[1].trim(), structured: true };
  }

  // Database ID + section: "hu-law-2012-1-00-00 s116" or "hu-law-2013-5-00-00 s6:272"
  const dbIdWithSection = trimmed.match(
    /^(hu-law-\d{4}-\d+-\d{2}-\d{2})\s+s?(\d+(?::\d+)?(?:\/[A-Za-z])?)$/i
  );
  if (dbIdWithSection) {
    return { documentRef: dbIdWithSection[1], sectionRef: dbIdWithSection[2], structured: true };
  }

  // Database ID only: "hu-law-2012-1-00-00"
  const dbIdOnly = trimmed.match(
    /^(hu-law-\d{4}-\d+-\d{2}-\d{2})$/
  );
  if (dbIdOnly) {
    return { documentRef: dbIdOnly[1], structured: true };
  }

  // "Section N <Act>" or "Section N, <Act>"
  const sectionFirst = trimmed.match(SECTION_FIRST_RE);
  if (sectionFirst) {
    return { documentRef: sectionFirst[2].trim(), sectionRef: sectionFirst[1], structured: false };
  }

  // "<Act> s N" / "<Act>, s N" / "<Act> s. N" or "<Act> Section N" / "<Act>, Section N"
  const sectionLast = trimmed.match(SECTION_LAST_RE);
  if (sectionLast) {
    return { documentRef: sectionLast[1].trim(), sectionRef: sectionLast[2], structured: false };
  }

  // Just a document reference (no section)
  return { documentRef: trimmed, structured: false };
}

export async function validateCitationTool(
  db: InstanceType<typeof Database>,
  input: ValidateCitationInput,
): Promise<ToolResponse<ValidateCitationResult>> {
  const warnings: string[] = [];
  const parsed = parseCitation(input.citation);

  if (!parsed) {
    return {
      results: {
        valid: false,
        citation: input.citation,
        warnings: ['Could not parse citation format'],
      },
      _metadata: generateResponseMetadata(db),
    };
  }

  const docId = resolveDocumentId(db, parsed.documentRef);
  if (!docId) {
    return {
      results: {
        valid: false,
        citation: input.citation,
        warnings: [`Document not found: "${parsed.documentRef}"`],
      },
      _metadata: generateResponseMetadata(db),
    };
  }

  const doc = db.prepare(
    'SELECT id, title, status FROM legal_documents WHERE id = ?'
  ).get(docId) as { id: string; title: string; status: string };

  if (doc.status === 'repealed') {
    warnings.push(`WARNING: This statute has been repealed.`);
  } else if (doc.status === 'amended') {
    warnings.push(`Note: This statute has been amended. Verify you are referencing the current version.`);
  }

  if (parsed.sectionRef) {
    // Normalize section ref: "6:272" → try "s6272", "s6:272", "6:272", "6272"
    const sectionClean = parsed.sectionRef.replace(':', '');
    const provision = db.prepare(
      "SELECT provision_ref FROM legal_provisions WHERE document_id = ? AND (provision_ref = ? OR provision_ref = ? OR provision_ref = ? OR provision_ref = ? OR section = ? OR section = ?)"
    ).get(docId, parsed.sectionRef, `s${parsed.sectionRef}`, `s${sectionClean}`, sectionClean, parsed.sectionRef, sectionClean) as { provision_ref: string } | undefined;

    if (!provision) {
      return {
        results: {
          valid: false,
          citation: input.citation,
          document_id: docId,
          document_title: doc.title,
          warnings: [...warnings, `Provision "${parsed.sectionRef}. §" not found in ${doc.title}`],
        },
        _metadata: generateResponseMetadata(db),
      };
    }

    return {
      results: {
        valid: true,
        citation: input.citation,
        normalized: `${doc.title} ${parsed.sectionRef}. § (Section ${parsed.sectionRef})`,
        document_id: docId,
        document_title: doc.title,
        provision_ref: provision.provision_ref,
        status: doc.status,
        warnings,
      },
      _metadata: generateResponseMetadata(db),
    };
  }

  return {
    results: {
      valid: true,
      citation: input.citation,
      normalized: doc.title,
      document_id: docId,
      document_title: doc.title,
      status: doc.status,
      warnings,
    },
    _metadata: generateResponseMetadata(db),
  };
}

File: internal/tools/validate_citation.go
// validate_citation — Validate an Hungarian legal citation against the
// database. Port of src/tools/validate-citation.ts. ParseCitation is exported
// and reused by format_citation, exactly like in TypeScript.

package tools

import (
	"context"
	"database/sql"
	"errors"
	"fmt"
	"regexp"
	"strings"

	"github.com/ryzen3100/magyar-jogszabaly-mcp/v2/internal/statute"
)

// validateCitationArgs mirrors ValidateCitationInput.
type validateCitationArgs struct {
	Citation *string `json:"citation"`
}

// ValidateCitationResult — singular object result. normalized/document_id/
// document_title/provision_ref/status are omitted when unset (the TS
// interface's optional fields dropped by JSON.stringify). Field order matches
// the TS interface.
type ValidateCitationResult struct {
	Valid         bool     `json:"valid"`
	Citation      string   `json:"citation"`
	Normalized    string   `json:"normalized,omitempty"`
	DocumentID    string   `json:"document_id,omitempty"`
	DocumentTitle string   `json:"document_title,omitempty"`
	ProvisionRef  string   `json:"provision_ref,omitempty"`
	Status        string   `json:"status,omitempty"`
	Warnings      []string `json:"warnings"`
}

// ParsedCitation is the parse outcome; SectionRef is "" when the citation
// carries no section.
type ParsedCitation struct {
	DocumentRef string
	SectionRef  string
	// Structured is true when DocumentRef is a formal Hungarian reference or
	// a database ID.
	Structured bool
}

// Citation grammars, ported regex for regex from validate-citation.ts.
// All are RE2-safe; (?i) matches the JS /i flag.
var (
	hungarianFullRe = regexp.MustCompile(`(?i)^(\d{4}\.\s*évi\s+[IVXLCDM]+\.\s*törvény)\s+` +
		`(\d+(?::\d+)?(?:\/[A-Za-z])?)\.\s*§`)
	hungarianDocRe = regexp.MustCompile(`(?i)^(\d{4}\.\s*évi\s+[IVXLCDM]+\.\s*törvény)$`)
	// Decree + section: "210/2009. Korm. rendelet 1. §",
	// "210/2009. (IX. 29.) Korm. rendelet 1. §" and the full-title form
	// "…rendelet a kereskedelmi tevékenységek végzésének feltételeiről 1. §".
	// Greedy group 1 keeps everything up to the LAST trailing "N. §" (decree
	// titles are long and carry the promulgation date); the section grammar
	// matches hungarianFullRe.
	decreeFullRe   = regexp.MustCompile(`(?i)^(\d{1,4}/\d{4}\.?.*rendelet.*)\s+(\d+(?::\d+)?(?:\/[A-Za-z])?)\s*\.?\s*§$`)
	dbIDWithSecRe  = regexp.MustCompile(`(?i)^(hu-law-\d{4}-\d+-\d{2}-\d{2})\s+s?(\d+(?::\d+)?(?:\/[A-Za-z])?)$`)
	dbIDOnlyRe     = regexp.MustCompile(`^(hu-law-\d{4}-\d+-\d{2}-\d{2})$`)
	sectionFirstRe = regexp.MustCompile(`(?i)^Section\s+(\d+[A-Za-z]*(?:\(\d+\))?)\s*[,;]?\s+(.+)$`)
	sectionLastRe  = regexp.MustCompile(`(?i)^(.+?)\s*[,;]?\s+(?:s\.?\s+|Section\s+)(\d+[A-Za-z]*(?:\(\d+\))?)$`)
)

// ParseCitation parses an Hungarian legal citation. Returns nil only for an
// empty/whitespace string (the TS null).
func ParseCitation(citation string) *ParsedCitation {
	trimmed := strings.TrimSpace(citation)
	if trimmed == "" {
		return nil
	}

	// Hungarian formal: "2012. évi I. törvény 116. §" / "6:272. §" / "116/A. §"
	if m := hungarianFullRe.FindStringSubmatch(trimmed); m != nil {
		return &ParsedCitation{DocumentRef: strings.TrimSpace(m[1]), SectionRef: m[2], Structured: true}
	}

	// Hungarian document only: "2012. évi I. törvény" (no section)
	if m := hungarianDocRe.FindStringSubmatch(trimmed); m != nil {
		return &ParsedCitation{DocumentRef: strings.TrimSpace(m[1]), Structured: true}
	}

	// Decree + section: "210/2009. Korm. rendelet 1. §"
	if m := decreeFullRe.FindStringSubmatch(trimmed); m != nil {
		return &ParsedCitation{DocumentRef: strings.TrimSpace(m[1]), SectionRef: m[2], Structured: true}
	}

	// Database ID + section: "hu-law-2012-1-00-00 s116" / "… s6:272"
	if m := dbIDWithSecRe.FindStringSubmatch(trimmed); m != nil {
		return &ParsedCitation{DocumentRef: m[1], SectionRef: m[2], Structured: true}
	}

	// Database ID only: "hu-law-2012-1-00-00"
	if m := dbIDOnlyRe.FindStringSubmatch(trimmed); m != nil {
		return &ParsedCitation{DocumentRef: m[1], Structured: true}
	}

	// "Section N <Act>" or "Section N, <Act>"
	if m := sectionFirstRe.FindStringSubmatch(trimmed); m != nil {
		return &ParsedCitation{DocumentRef: strings.TrimSpace(m[2]), SectionRef: m[1]}
	}

	// "<Act> s N" / "<Act>, s N" / "<Act> s. N" / "<Act> Section N"
	if m := sectionLastRe.FindStringSubmatch(trimmed); m != nil {
		return &ParsedCitation{DocumentRef: strings.TrimSpace(m[1]), SectionRef: m[2]}
	}

	// Just a document reference (no section)
	return &ParsedCitation{DocumentRef: trimmed}
}

// ValidateCitation is the exported handler for validate_citation. It always
// returns a singular result — unparseable citations and unknown documents
// come back as valid:false plus warnings, not errors — and _metadata.note
// stays unset.
func ValidateCitation(ctx context.Context, db *sql.DB, args map[string]any) (any, ResponseMetadata, error) {
	var parsed validateCitationArgs
	if err := decodeArgs(args, &parsed); err != nil {
		return nil, ResponseMetadata{}, err
	}
	if err := validateArgs(
		checkRequired("citation", parsed.Citation),
		checkMaxLength("citation", parsed.Citation, maxCitationLength),
	); err != nil {
		return nil, ResponseMetadata{}, err
	}

	warnings := []string{}
	parsedCitation := ParseCitation(*parsed.Citation)
	if parsedCitation == nil {
		return ValidateCitationResult{
			Valid:    false,
			Citation: *parsed.Citation,
			Warnings: []string{"Could not parse citation format"},
		}, GenerateResponseMetadata(ctx, db), nil
	}

	docID, err := statute.ResolveDocumentID(ctx, db, parsedCitation.DocumentRef)
	if err != nil {
		return nil, ResponseMetadata{}, fmt.Errorf("resolve document: %w", err)
	}
	if docID == "" {
		return ValidateCitationResult{
			Valid:    false,
			Citation: *parsed.Citation,
			Warnings: []string{fmt.Sprintf("Document not found: \"%s\"", parsedCitation.DocumentRef)},
		}, GenerateResponseMetadata(ctx, db), nil
	}

	var title, status string
	err = db.QueryRowContext(ctx, "SELECT title, status FROM legal_documents WHERE id = ?",
		docID).Scan(&title, &status)
	if err != nil {
		return nil, ResponseMetadata{}, fmt.Errorf("query document: %w", err)
	}

	switch status {
	case "repealed":
		warnings = append(warnings, "WARNING: This statute has been repealed.")
	case "amended":
		warnings = append(warnings, "Note: This statute has been amended. Verify you are referencing the current version.")
	}

	if parsedCitation.SectionRef != "" {
		// Same normalized exact-match candidates as get_provision (shared
		// tolerance — audit E5): "6:272" → section "6:272"/"6272",
		// provision_ref "s6272".
		candidates := statute.SectionRefCandidates(parsedCitation.SectionRef)
		var provisionRef string
		err := sql.ErrNoRows
		if len(candidates) > 0 {
			where, args := provisionRefQuery(docID, candidates)
			err = db.QueryRowContext(ctx,
				"SELECT provision_ref FROM legal_provisions WHERE "+where+" ORDER BY id LIMIT 1",
				args...,
			).Scan(&provisionRef)
		}
		if errors.Is(err, sql.ErrNoRows) {
			warnings = append(warnings, fmt.Sprintf("Provision \"%s. §\" not found in %s", parsedCitation.SectionRef, title))
			return ValidateCitationResult{
				Valid:         false,
				Citation:      *parsed.Citation,
				DocumentID:    docID,
				DocumentTitle: title,
				Warnings:      warnings,
			}, GenerateResponseMetadata(ctx, db), nil
		}
		if err != nil {
			return nil, ResponseMetadata{}, fmt.Errorf("query provision: %w", err)
		}

		return ValidateCitationResult{
			Valid:         true,
			Citation:      *parsed.Citation,
			Normalized:    fmt.Sprintf("%s %s. § (Section %s)", title, parsedCitation.SectionRef, parsedCitation.SectionRef),
			DocumentID:    docID,
			DocumentTitle: title,
			ProvisionRef:  provisionRef,
			Status:        status,
			Warnings:      warnings,
		}, GenerateResponseMetadata(ctx, db), nil
	}

	return ValidateCitationResult{
		Valid:         true,
		Citation:      *parsed.Citation,
		Normalized:    title,
		DocumentID:    docID,
		DocumentTitle: title,
		Status:        status,
		Warnings:      warnings,
	}, GenerateResponseMetadata(ctx, db), nil
}
